beps/bep_0056.rst

```diff
@@ -0,0 +1,145 @@
+
+
+
+The compression algorithm is selected by taking the dictionary item with
+highest priority from intersection of items supported by both peers,
+if there isn't any suitable compression algorithm - compression will be disabled.
+
+Example of extension handshake message:
+
+::
+
+  {
+    'c': {
+      'p_zstd': 255,
+      's_zstd': 153,
+      'p_lz4': 106,
+      'p_density': 70,
+      's_lz4': 41,
+      's_density': 37
+    }
+  }
+
+
+Compression methods
+===================
+Extension provides two approaches (methods) to compression, which have
+their own trade-offs, so choice between these should be made by clients
+on per-torrent basis, using its metadata (properties like piece size).
+
+With **by-piece compression** method, client must compress each piece
+individually, which lowers overall compression ratio but result can
+be stored in cache and reused, probably providing more efficiency.
+If the client is caching compressed pieces in memory, then it can be
+decompressed when saving to disk or sending to peer, which not supports
+compression. To reduce piece re-compression, client should raise
+current algorithm's priority during handshake. This method has low
+efficiency with pieces smaller than 4 MB.
+
+Clients using **stream compression** method instead compresses whole
+data stream, so compression ratio should be higher. During handshake,
+clients should lower or raise algorithm's priority depending on expected
+factors that could impact compression efficiency and performance. This
+method can introduce performance issues if used on thousands of
+simultaneous connections.
+
+Allowed compression algorithms
+------------------------------
+
+Compression algorithms must satisfy the following requirements:
+
+1. Decompression speed must not be lower than 500 MB/s.
+
+2. It must not produce a larger piece than the original by 1%.
+
+For consistency, identifiers are prefixed by ``p_`` or ``s_``
+for "piece" and "stream" compression methods accordingly.
+
++-------------+-----------------------------+
+| identifier  | compression algorithm       |
++=============+=============================+
+| p_lz4       | LZ4                         |
++-------------+-----------------------------+
+| s_lz4       | LZ4                         |
++-------------+-----------------------------+
+| p_density   | Chameleon (DENSITY library) |
++-------------+-----------------------------+
+| s_density   | Chameleon (DENSITY library) |
++-------------+-----------------------------+
+| p_zstd      | ZStandard                   |
++-------------+-----------------------------+
+| s_zstd      | ZStandard                   |
++-------------+-----------------------------+
+
+This specification deliberately doesn't provide negotiation
+for configuration options, default ones must be used unless
+specified otherwise.
+
+**NOTE**: Currently, only ``p_zstd`` and ``s_zstd`` algorithms
+are required for implementation.
```

Comment:
What's the point of requiring this? Would it be a problem if the negotiations resulted in an empty set of algorithms and normal protocol was used?

Reply:
There were concerns about different clients supporting non-overlapping sets of algorithms, so specification should require one that must be implemented universally. There wouldn't be a problem if negotiations resulted in an empty set, as the compression feature could be disabled by the user.

Reply:
this is an extension to begin with. There will be clients not implementing it. I don't see a problem with that.

Reply:
Currently, algorithm list must be reworked, as there can be additional options.
Until it's necessary, I removed the note.